packages/abi/src/abi-type.ts

```diff
@@ -1,718 +1,904 @@
+ *
+ * This is the abstract base class for all ABI types.
  */
-export function getABITypeName(abiType: ABIType): string {
-  switch (abiType.name) {
-    case ABITypeName.Uint:
-      return uintToString(abiType)
-    case ABITypeName.Ufixed:
-      return ufixedToString(abiType)
-    case ABITypeName.Address:
-      return 'address'
-    case ABITypeName.Bool:
-      return 'bool'
-    case ABITypeName.Byte:
-      return 'byte'
-    case ABITypeName.String:
-      return 'string'
-    case ABITypeName.Tuple:
-      return tupleToString(abiType)
-    case ABITypeName.StaticArray:
-      return staticArrayToString(abiType)
-    case ABITypeName.DynamicArray:
-      return dynamicArrayToString(abiType)
-    case ABITypeName.Struct:
-      return structToString(abiType)
+export abstract class ABIType {
+  /**
+   * Returns the ARC-4 type name string representation.
+   * @returns The ARC-4 type string
+   */
+  abstract get name(): string
+
+  /**
+   * Returns a user-friendly display name for this type.
+   * @returns The display name for this type
+   */
+  get displayName(): string {
+    return this.name
+  }
+
+  /**
+   * Returns the ARC-4 type name string representation.
+   * @returns The ARC-4 type string
+   */
+  toString(): string {
+    return this.name
+  }
+
+  /**
+   * Checks if this ABI type is equal to another.
+   * @param other The other ABI type to compare with
+   * @returns True if the types are equal, false otherwise
+   */
+  abstract equals(other: ABIType): boolean
+
+  /**
+   * Checks if this ABI type is dynamic (variable-length).
+   * @returns True if the type is dynamic, false otherwise
+   */
+  abstract isDynamic(): boolean
+
+  /**
+   * Gets the byte length of the encoded type for static types.
+   * @returns The number of bytes needed to encode this type
+   * @throws Error if the type is dynamic
+   */
+  abstract byteLen(): number
+
+  /**
+   * Encodes a value according to this ABI type.
+   * @param value The value to encode
+   * @returns The encoded bytes
+   */
+  abstract encode(value: ABIValue): Uint8Array
+
+  /**
+   * Decodes bytes according to this ABI type.
+   * @param bytes The bytes to decode
+   * @returns The decoded value
+   */
+  abstract decode(bytes: Uint8Array): ABIValue
+
+  /**
+   * Creates an ABI type from an ARC-4 type string.
+   * @param str The ARC-4 type string (e.g., "uint256", "bool", "(uint8,address)")
+   * @returns The corresponding ABI type
+   */
+  static from(str: string): ABIType {
+    if (str.endsWith('[]')) {
+      const childType = ABIType.from(str.slice(0, str.length - 2))
+      return new ABIArrayDynamicType(childType)
+    }
+    if (str.endsWith(']')) {
+      const stringMatches = str.match(STATIC_ARRAY_REGEX)
+      if (!stringMatches || stringMatches.length !== 3) {
+        throw new Error(`malformed static array string: ${str}`)
```

Comment:
```suggestion
        throw new Error(`Malformed static array string: ${str}`)
```

We tend to start error message with uppercase. This applies to all the changes removing the prefix.